crates/kubelet/src/volumes.rs

```diff
@@ -0,0 +1,186 @@
+}
+
+/// A smart wrapper around the location of a volume on the host system. If this is a ConfigMap or
+/// Secret volume, dropping this reference will clean up the temporary volume. [AsRef] and
+/// [std::ops::Deref] are implemented for this type so you can still use it like a normal PathBuf
+pub struct VolumeRef {
+    host_path: PathBuf,
+    volume_type: VolumeType,
+}
+
+impl VolumeRef {
+    /// Resolves the volumes for a pod, including preparing temporary directories containing the
+    /// contents of secrets and configmaps. Returns a HashMap of volume names to a PathBuf for the
+    /// directory where the volume is mounted
+    pub async fn volumes_from_pod(
+        volume_dir: &PathBuf,
+        pod: &Pod,
+        client: &kube::Client,
+    ) -> anyhow::Result<HashMap<String, Self>> {
+        let base_path = volume_dir.join(pod_dir_name(pod));
+        tokio::fs::create_dir_all(&base_path).await?;
+        if let Some(vols) = pod.volumes() {
+            let volumes = vols.iter().map(|v| {
+                let mut host_path = base_path.clone();
+                host_path.push(&v.name);
+                async move {
+                    let volume_type = configure(v, pod.namespace(), client, &host_path).await?;
+                    Ok((
+                        v.name.to_owned(),
+                        VolumeRef {
+                            host_path,
+                            volume_type,
+                        },
+                    ))
+                }
+            });
+            futures::future::join_all(volumes)
+                .await
+                .into_iter()
+                .collect()
+        } else {
+            Ok(HashMap::default())
+        }
+    }
+}
+
+impl AsRef<PathBuf> for VolumeRef {
+    fn as_ref(&self) -> &PathBuf {
+        &self.host_path
+    }
+}
+
+impl Deref for VolumeRef {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.host_path
+    }
+}
+
+impl Drop for VolumeRef {
+    fn drop(&mut self) {
+        if matches!(self.volume_type, VolumeType::ConfigMap | VolumeType::Secret) {
+            // TODO: Currently there is no way to do this async (though there is an async destructors proposal)
+            debug!(
+                "deleting {:?} directory {:?}",
+                self.volume_type, self.host_path
+            );
+            std::fs::remove_dir_all(&self.host_path).unwrap_or_else(|e| {
+                error!(
+                    "unable to delete directory {:?} on volume cleanup: {:?}",
+                    self.host_path, e
+                )
+            });
+        }
+    }
+}
+
+/// This is a gnarly function to check all of the supported data members of the Volume struct.
+/// Because it isn't a HashMap, we need to check all fields individually
+async fn configure(
+    vol: &KubeVolume,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    if let Some(cm) = &vol.config_map {
```

Comment:
nit I think this would read better as match instead of a long `if else` chain

Reply:
I'm also wondering if there's a way the `secret_client` or `configmap_client` could be part of the KubeVolume, kind of like a storage driver works for Helm. That way a caller cannot place a KubeVolume in a location other than where the creator intended the volume to go. That would clean up `populate` to just take a `path: &PathBuf`. That being said, this is third-party code so it might not be possible without going through a few hoops, so I'll let you make the call on that.

That cleanup could potentially make it worthwhile implementing `.populate` as a trait for a KubeVolume. That way we can just call `vol.populate(path)` and leave the KubeVolume to error out if it does not support `populate`. Then again, I might be misunderstanding how traits work in Rust so 🤷‍♂️ 

What are your thoughts?

Reply:
@bacongobbler let me look at it again

@rylev I tried to do a match but I couldn't get it to work as these are all individual data members of a struct. Is there another way to do this?

Reply:
@rylev and I talked offline and decided it would be better this way